### DISCUSSION TITLE
scriptlet not working as expected

### DISCUSSION THREAD

**Discussion by uBlock-user** [uBlockOrigin/uBlock-issues#2676] [Q&A]
@gorhill Add `deadline.com##+js(rmnt, script, /pmc/)` to my filters and browse to `https://deadline.com/`

In Nightly, press F12 and open dev-tools, upon opening, browse to Inspector pane and if you search for `pmc`, devtools will hang/lag and will find many instances of `pmc` in `<script>` tags even though there should be none(essentially scriptlet has failed to remove pmc ladden js)

In Firefox, repeat the same procedure above, and dev-tools will not hang but scriptlet doesn't work either, evidenced by presense of pmc ladden js as per the searches in Inspector pane of devtools. 

Are you seeing this behaviour ?

**Comment by gorhill** [comment]
It's full of instances of `pmc` in there outside `script` tags. When using:

    deadline.com##+js(rmnt, script, /pmc/, log, 1)

Then reload the page, I see lot of reports of `script` tags being emptied, as expected.

---

Note that I get a lot errors such as `Uncaught ReferenceError: pmc_breaking_news_hash is not defined` or `Uncaught ReferenceError: pmcRecaptcha is not defined`, with that filter, suggesting it did prevent at least some inline JS from executing.

**Reply by mapx-** [reply]
It's not like in FF with html filtering when pmc will disappear from the page code.

**Reply by uBlock-user** [reply]
It's not disappearing from Inspector pane, see -- 
![Capture](https://github.com/uBlockOrigin/uBlock-issues/assets/21290713/19f6322a-c051-49e1-b2b8-e7f38f83e427)

**Reply by uBlock-user** [reply]
Secondly, why does devtools experience such lag/hang in Nightly ?

**Reply by gorhill** [reply]
> Secondly, why does devtools experience such lag/hang in Nightly ?

It doesn't on my side.

**Reply by uBlock-user** [reply]
> Note that I get a lot errors such as Uncaught ReferenceError: pmc_breaking_news_hash is not defined or Uncaught ReferenceError: pmcRecaptcha is not defined, with that filter, suggesting it did prevent at least some inline JS from executing.

Yes, seems this filter will be partially successful only based on foiling few inline JS.

**Reply by gorhill** [reply]
> It's not disappearing from Inspector pane, see --

That script is reported as being replaced on my side in the dev console, and when I try to find `window.pmc` in the inspector, it's not found. It's possible the scriptlet is injected later in your case for whatever reason, maybe you have other extensions delaying the injection time.

**Reply by uBlock-user** [reply]
> when I try to find window.pmc in the inspector, it's not found.

On my side, it finds it. I don't have any extensions installed besides uBlock Origin dev build. Just try this and let me know what you see --

In the inspector pane, search for "pmc "(space is intentional so you don't get trapped with the selector results, remove the quotes too) and post the number of results appear here. It should be below 8 with the filter.

**Reply by gorhill** [reply]
> On my side, it finds it.

On my side it finds 2, both of which are inside HTML comments.

Was it reported as removed once you use `log, 1`?

As this was discussed in the past, ability to foil inline JS when removing text in mutation observer is a cool trick but it's just a side-effect of some obscure specs, by no mean there is a guarantee this will work, especially more so if for some reasons the scriptlet is injected too late. It's not something I can "fix", and to figure why this happens on your side would involve investigating your specific setup, not something I can do. If it does not work, use HTML filtering, that one work in a deterministic way.

**Reply by uBlock-user** [reply]
> Was it reported as removed once you use log, 1?

It does show before and after(empty string) in the log in console. 


> ability to foil inline JS when removing text in mutation observer is a cool trick but it's just a side-effect of some obscure specs, by no mean there is a guarantee this will work, especially more so if for some reasons the scriptlet is injected too late. 

Well my own [scriptlet](https://github.com/uBlock-user/uBO-Scriptlets/blob/master/scriptlets.js#L39) works using MO and inspector pane doesn't show any script containing a string with pmc. With the exception of race condition and that too on some sites, it always works.

**Reply by gorhill** [reply]
It works on my side, don't know what else to say. Since you have no issue with your own script, maybe just tell me what needs to be done to make it work on your side.

**Reply by uBlock-user** [reply]
Maybe for rmnt, also add [remove()](https://developer.mozilla.org/en-US/docs/Web/API/Element/remove) after replacing `textContent` to ''' ? This is something you did for procedural operator `:remove()` too.

**Reply by gorhill** [reply]
You didn't report on what `log, 1` shows on your side. Before making any changes, you need to investigate more on your side. A `remove()` wouldn't accomplish anything if the inline script has been executed already (for example uBO itself does this to its injected inline script and that doesn't prevent execution of these).

**Reply by uBlock-user** [reply]
> You didn't report on what log, 1 

Yes, I did. See below.

 > It does show before and after(empty string) in the log in console.

image - ![Capture](https://github.com/uBlockOrigin/uBlock-issues/assets/21290713/938c62b5-95b2-4b6f-8a72-cb7c7dc33b51)


---


> Before making any changes, you need to investigate more on your side.

I wouldn't know where to investigate. Anyways, given it works for you, so it must be timing issue on my side.

**Reply by gorhill** [reply]
Do you have an entry which says something like `x nodes present before installing mutation observer`?

**Reply by uBlock-user** [reply]
Yes, sometimes this appears - ![Capture](https://github.com/uBlockOrigin/uBlock-issues/assets/21290713/f763f440-91a9-4c99-b785-8dc2b064e5b9)

**Reply by gorhill** [reply]
Sorry, at first I tested on Firefox stable, but then moved to testing onto Nightly which is my main browser as I thought this could be a Nightly-specific issue, and forgot to enable JS. Now when I test in a blank profile in Firefox stable, I can find `window.pmc.analytics` in the inspector despite being reportedly removed. The debugger shows those lines as dimmed, i.e. can't put a breakpoint on these, but then typing `window.pmc.analytics` in the dev console shows that it is an `Array`. Adding a `node.remove()` when `node.nodeType === 1` makes no difference. So essentially we can't reliably prevent all inline scripts from being executed with the `rmnt` trick, and it might be related to the fact the that those inline scripts are at the bottom of the of the document -- if I remember correctly the script execution is being held back waiting for something related to stylesheets. Now if you say it works with your custom scriptlet, I have no explanation.

---

Correction: 

> typing `window.pmc.analytics` in the dev console shows that it is an `Array`

That's not a good test, as this could be set by a 3rd-party script. A better test is `global_urlhashchanged`, which is a global function in that inline script, and it's reported as not defined in the dev console. So the inline script has been defused after all.

**Reply by uBlock-user** [reply]
>  Now if you say it works with your custom scriptlet, I have no explanation.

Yes, what I mean by, specifically is, inline scripts get removed from the DOM tree present shown in the the Inspector pane. That's all.

**Reply by gorhill** [reply]
An important difference is that your custom scriptlet is executed in page world through a blob URL, i.e. probably quite later than `rmnt`.

**Reply by uBlock-user** [reply]
> i.e. probably quite later than

Any way I can make it execute in 'isolated' world like your scriptlet ?

**Reply by gorhill** [reply]
I would need to add support in the parser for this. What is the URL of your scriptlet?

**Reply by uBlock-user** [reply]
https://ubo-scriptlets.pages.dev/scriptlets.js ?

**Reply by gorhill** [reply]
I've made changes to support `world` and `dependency`, <https://github.com/gorhill/uBlock/commit/695b93f9708a>, I feel it's a safe change to release candidate since this code is not used in normal uBO installation. Usage example for your case:

    /// remove-node.js
    /// alias rn.js
    /// world isolated

I also added `dependency` if you want to use some of the `.fn` helpers in <https://github.com/gorhill/uBlock/blob/master/assets/resources/scriptlets.js>. Example:

    /// remove-node.js
    /// alias rn.js
    /// world isolated
    /// dependency pattern-to-regex.fn

The helpers are for internal use first and foremost, and it may happen I change them whenever I feel like doing so, so I offer no guarantee that they will always work as they currently do. You can declare as many `dependency` as needed.

**Reply by uBlock-user** [reply]
I couldn't find any list of advantages of using "ISOLATED" over "MAIN" world. Do you mind listing them here ?

**Reply by gorhill** [reply]
Steps:

- MAIN
    - Firefox:
        - Content script code assembled in background process
        - Content script code injected from background process
        - Content script code create `blob:` URL containing all scriptlets for the site
        - Content script code add `script` tag to the DOM with `blob:` URL as source
        - Scriptlet code executes at some later point when the browser decides to do so
    - Chromium:
        - Content script code assembled in background process
        - Content script code injected from background process
        - Content script code create inline `script` tag containing all scriptlets as text content
        - Content script code add inline `script` tag to the DOM
        - Scriptlet code executes immediately upon insertion
- ISOLATED
    - Firefox, Chromium:
        - Content script code assembled in background process
        - Content script code injected from background process
        - Content script code contains scriptlet code, so they execute immediately, no need to insert anything in DOM

MAIN: can access page's variables, ISOLATED can't.
